File: src/leetcode/editor/cn/[7]整数反转.java
package leetcode.editor.cn;
//给你一个 32 位的有符号整数 x ，返回 x 中每位上的数字反转后的结果。 
//
// 如果反转后整数超过 32 位的有符号整数的范围 [−231, 231 − 1] ，就返回 0。 
//假设环境不允许存储 64 位整数（有符号或无符号）。
//
// 
//
// 示例 1： 
//
// 
//输入：x = 123
//输出：321
// 
//
// 示例 2： 
//
// 
//输入：x = -123
//输出：-321
// 
//
// 示例 3： 
//
// 
//输入：x = 120
//输出：21
// 
//
// 示例 4： 
//
// 
//输入：x = 0
//输出：0
// 
//
// 
//
// 提示： 
//
// 
// -231 <= x <= 231 - 1 
// 
// Related Topics 数学 
// 👍 2523 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Solution7 {
    //题解 https://leetcode-cn.com/problems/reverse-integer/solution/tu-jie-7-zheng-shu-fan-zhuan-by-wang_ni_ma/
    //栈、取模取余 或者 变成字符串反转但是要注意符号
    //这里使用取模解法
    public int reverse(int x) {
        int res = 0;
        while (x != 0) {
            //取尾巴
            int temp = x % 10;
            //判断最后一位是否溢出
            if (res > Integer.MAX_VALUE / 10 || (res == Integer.MAX_VALUE / 10 && temp > Integer.MAX_VALUE % 10)) {
                return 0;
            }
            if (res < Integer.MIN_VALUE / 10 || (res == Integer.MIN_VALUE / 10 && temp < Integer.MIN_VALUE % 10)) {
                return 0;
            }
            res = res * 10 + temp;
            x /= 10;

        }
        return res;


    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: src/leetcode/editor/cn/[7]整数反转.kt
//给出一个 32 位的有符号整数，你需要将这个整数中每位上的数字进行反转。 
//
// 示例 1: 
//
// 输入: 123
//输出: 321
// 
//
// 示例 2: 
//
// 输入: -123
//输出: -321
// 
//
// 示例 3: 
//
// 输入: 120
//输出: 21
// 
//
// 注意: 
//
// 假设我们的环境只能存储得下 32 位的有符号整数，则其数值范围为 [−231, 231 − 1]。请根据这个假设，如果反转后整数溢出那么就返回 0。 
// Related Topics 数学

package leetcode.editor.cn

object ReverseInteger {
    @JvmStatic
    fun main(args: Array<String>) {
        var solution: Solution = Solution()
        print(solution.reverse(123))
    }


    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        fun reverse(x: Int): Int {
            var xin = x
            var reverseReslut: Long = 0
            while (xin != 0) {
                reverseReslut = reverseReslut * 10 + xin % 10
                if (reverseReslut > Int.MAX_VALUE || reverseReslut < Int.MIN_VALUE) {
                    return 0
                }
                xin /= 10
            }

            return reverseReslut.toInt()

        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}